src/suit_decoder.c

```diff
@@ -0,0 +1,822 @@
+
+			/* If authentication for any key fails, drop the envelope */
+			if (ret != SUIT_SUCCESS) {
+				results[i * 2] = suit_bool_false;
+			}
+
+			if (ret == SUIT_SUCCESS) {
+				results[i * 2] = suit_bool_true;
+			}
+		}
+
+		for (i = 0; i < state->authentication_bstr_count; i++) {
+			if (results[i * 2 + 1] != suit_bool_false) {
+				ret = SUIT_ERR_TAMP;
+			}
+
+			if (results[i * 2] == suit_bool_true) {
+				if (num_ok == i) {
+					num_ok++;
+				}
+			}
+		}
+
+		if (num_ok == state->authentication_bstr_count) {
+			ret = SUIT_SUCCESS;
+		} else if (ret != SUIT_ERR_TAMP) {
+			ret = SUIT_ERR_AUTHENTICATION;
+		}
+	}
+
+	if (ret == ZCBOR_SUCCESS) {
+		state->step = MANIFEST_AUTHENTICATED;
+		return SUIT_SUCCESS;
+	}
+
+	suit_decoder_reset_state(state);
+	if (ret != SUIT_SUCCESS) {
+		return ret;
+	}
+
+	return SUIT_ERR_TAMP;
+}
+
+int suit_decoder_authorize_manifest(struct suit_decoder_state *state)
+{
+	struct SUIT_Common *common = NULL;
+	int ret = SUIT_SUCCESS;
+
+	if (state == NULL) {
+		return SUIT_ERR_DECODING;
+	}
+
+	if (state->step != MANIFEST_AUTHENTICATED) {
+		return SUIT_ERR_ORDER;
+	}
+
+	/* Verify manifest version - enforced by the CDDL and checked by the ZCBOR parser code */
+
+	/* Cannot perform universal sequence number authorization.
+	 * Skip this check in decoder logic and store the sequence number value inside the output structure.
+	 */
+	state->decoded_manifest->sequence_number = state->manifest._SUIT_Manifest_suit_manifest_sequence_number;
+
+	/* Verify common sequence */
+	common = &state->manifest._SUIT_Manifest_suit_common_cbor;
+
+	/* Verify list of components */
+	if (common->_SUIT_Common_suit_components_present) {
+		struct zcbor_string component_id;
+
+		/* Verify the length of the list */
+		if ((common->_SUIT_Common_suit_components._SUIT_Common_suit_components._SUIT_Components__SUIT_Component_Identifier_count > SUIT_MAX_NUM_COMPONENTS) ||
+		    (common->_SUIT_Common_suit_components._SUIT_Common_suit_components._SUIT_Components__SUIT_Component_Identifier_count < 1)) {
+			return SUIT_ERR_MANIFEST_VALIDATION;
+		}
+
+		/* Authorize component IDs */
+		for (int i = 0; i < common->_SUIT_Common_suit_components._SUIT_Common_suit_components._SUIT_Components__SUIT_Component_Identifier_count; i++) {
+			/* Zip list of strings into a single ZCBOR string */
+			ret = get_component_id_str(&component_id,
+				&common->_SUIT_Common_suit_components._SUIT_Common_suit_components._SUIT_Components__SUIT_Component_Identifier[i]);
+
+#ifndef SUIT_PLATFORM_LEGACY_API_SUPPORT
+			if (ret == SUIT_SUCCESS) {
+				ret = suit_plat_authorize_component_id(&state->decoded_manifest->manifest_component_id, &component_id);
+			}
+#endif /* !SUIT_PLATFORM_LEGACY_API_SUPPORT */
+
+			if (ret == SUIT_SUCCESS) {
+				continue;
+			}
+
+			break;
+		}
+
+		/* Verify the depenecy indexes. */
+		if (ret == SUIT_SUCCESS) {
+			ret = check_dependency_indexes(common);
+			if (ret == SUIT_ERR_MISSING_COMPONENT) {
+				ret = SUIT_SUCCESS;
```

Comment:
Is that intentional? Returning SUCCESS in case of error?

Reply:
Yes, the `SUIT_ERR_MISSING_COMPONENT` error code indicates that the manifest does not contain dependencies.
It is not prohibited to have only regular components, so this situation does not invalidate manifest.
This error code was used, so the `check_dependency_indexes(..)` behavior matches with the `find_dependency_manifest_prefix(..)`.

Reply:
I think I will add a comment for that one. Maybe one day we will need to handle this case differently.